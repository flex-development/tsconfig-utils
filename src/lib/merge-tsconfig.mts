/**
 * @file mergeTsconfig
 * @module tsconfig-utils/lib/mergeTsconfig
 */

import getTsconfig from '#internal/get-tsconfig'
import type {
  JsonValue,
  Tsconfig,
  TsconfigHost
} from '@flex-development/tsconfig-utils'
import { isObjectCurly, properties } from '@flex-development/tutils'

/**
 * Merge one or more tsconfig objects into a single {@linkcode Tsconfig}.
 *
 * Tsconfig source objects are applied from left to right. Subsequent sources
 * overwrite property assignments of previous sources.
 *
 * > ðŸ‘‰ **Note**: If `target` is a {@linkcode TsconfigHost}, `target.tsconfig`
 * > will be modified. Otherwise, `target` will be modified.
 *
 * @see {@linkcode TsconfigHost}
 * @see {@linkcode Tsconfig}
 *
 * @this {void}
 *
 * @param {Tsconfig | TsconfigHost | null | undefined} target
 *  Target tsconfig or tsconfig host
 * @param {ReadonlyArray<Tsconfig | TsconfigHost | null | undefined>} tsconfigs
 *  Source tsconfig object(s)
 * @return {Tsconfig}
 *  Merged tsconfig
 */
function mergeTsconfig(
  this: void,
  target: Tsconfig | TsconfigHost | null | undefined,
  ...tsconfigs: readonly (Tsconfig | TsconfigHost | null | undefined)[]
): Tsconfig {
  return tsconfigs.reduce<Tsconfig>((acc, tsconfig) => {
    return properties(tsconfig = getTsconfig(tsconfig)).reduce((
      target,
      property
    ) => {
      /**
       * Merge value.
       *
       * @var {JsonValue | undefined} value
       */
      let value: JsonValue | undefined = tsconfig[property]

      if (Object.prototype.hasOwnProperty.call(target, property)) {
        /**
         * Target object value.
         *
         * @const {JsonValue} targetValue
         */
        const targetValue: JsonValue = target[property] as JsonValue

        /**
         * Source object value.
         *
         * @const {JsonValue} sourceValue
         */
        const sourceValue: JsonValue = tsconfig[property] as JsonValue

        if (isObjectCurly(targetValue) && isObjectCurly(sourceValue)) {
          value = { ...targetValue, ...sourceValue }
        }
      }

      return Object.assign(target, { [property]: value }), target
    }, acc)
  }, getTsconfig(target))
}

export default mergeTsconfig
