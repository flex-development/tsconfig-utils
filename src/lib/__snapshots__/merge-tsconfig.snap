// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`unit:lib/mergeTsconfig > should return merged tsconfig (target:Tsconfig) 1`] = `
{
  "compilerOptions": {
    "allowImportingTsExtensions": true,
    "allowJs": true,
    "alwaysStrict": true,
    "customConditions": [
      "tsconfig-utils",
      "development",
    ],
    "declaration": true,
    "declarationMap": true,
    "lib": [
      "es2023",
    ],
    "module": "esnext",
    "moduleResolution": "bundler",
    "newLine": "lf",
    "noEmit": true,
    "noErrorTruncation": true,
    "noUncheckedSideEffectImports": true,
    "noUnusedParameters": false,
    "paths": {
      "https://esm.sh/lodash@4.17.21": [
        "node_modules/@types/lodash/index.d.ts",
      ],
    },
    "resolveJsonModule": true,
    "sourceMap": true,
    "target": "esnext",
    "verbatimModuleSyntax": true,
  },
  "exclude": [
    "coverage",
    "dist",
    "node_modules",
  ],
  "extends": "../tsconfig.json",
  "include": [
    "**/**.json",
    "**/**.mjs",
    "**/**.mts",
    "**/**.ts",
  ],
  "mdx": {
    "plugins": [
      "remark-directive",
      "remark-frontmatter",
      "remark-gfm",
    ],
  },
}
`;

exports[`unit:lib/mergeTsconfig > should return merged tsconfig (target:TsconfigHost) 1`] = `
{
  "$schema": "https://json.schemastore.org/tsconfig",
  "_version": "2.0.0",
  "compilerOptions": {
    "allowImportingTsExtensions": true,
    "allowJs": true,
    "allowUnreachableCode": false,
    "allowUnusedLabels": false,
    "alwaysStrict": true,
    "checkJs": true,
    "customConditions": [
      "tsconfig-utils",
      "development",
    ],
    "declaration": true,
    "declarationMap": true,
    "esModuleInterop": true,
    "exactOptionalPropertyTypes": true,
    "isolatedModules": true,
    "lib": [
      "es2023",
    ],
    "module": "esnext",
    "moduleResolution": "bundler",
    "newLine": "lf",
    "noEmit": true,
    "noErrorTruncation": true,
    "noFallthroughCasesInSwitch": true,
    "noImplicitOverride": true,
    "noImplicitReturns": true,
    "noPropertyAccessFromIndexSignature": true,
    "noUncheckedIndexedAccess": true,
    "noUncheckedSideEffectImports": true,
    "noUnusedLocals": true,
    "noUnusedParameters": false,
    "paths": {
      "https://esm.sh/lodash@4.17.21": [
        "node_modules/@types/lodash/index.d.ts",
      ],
    },
    "resolveJsonModule": true,
    "skipLibCheck": true,
    "sourceMap": true,
    "strict": true,
    "target": "esnext",
    "verbatimModuleSyntax": true,
  },
  "exclude": [
    "coverage",
    "dist",
    "node_modules",
  ],
  "extends": "@tsconfig/strictest/tsconfig.json",
  "include": [
    "**/**.json",
    "**/**.mjs",
    "**/**.mts",
    "**/**.ts",
  ],
  "mdx": {
    "plugins": [
      "remark-directive",
      "remark-frontmatter",
      "remark-gfm",
    ],
  },
}
`;
